brave/brave-ios
# Fix #8205: Support NFT removal. Updates for NFT spam management new design

## Summary of Changes
User now can remove NFTs from the wallet, not just hide and unhide. 
Some new updates for NFT spam management please refer to the issue itself. 


This pull request fixes #8205

## Submitter Checklist:


## Test Plan:

1. Set up a wallet and go to wallet/portfolio/NFT
2. Check if the dropdown now has two options: 
    * Collected
    * Hidden
3. If you have auto-discovered NFTs and auto-discovery is enabled, check if 
    * auto-discovered NFTs are being displayed under `Collected`
    * any junk auto-discovered NFTs are being displayed under `Hidden`
    * check junk NFT has a `Junk` label on the top left.
4. Check if NFTs under `Collected` have context menu with options: 
    * Hide 
    * Don't show in wallet
5. Check clicking `Hide` in step 4 will move NFT to `Hidden`
6. Check clicking `Don't show in wallet` in step 4 will remove NFT which you can't see it in neither `Collected` nor `Hidden`
7. Check if NFTs that are not junk under `Hidden` have context menu with options:
    * Unhide
    * Don't show in wallet
8. Check clicking `Unhide` in step 7 will move NFT to `Collected`
9. Check clicking `Don't show in wallet` in step 7 will remove NFT which you can't see it in neither `Collected` nor `Hidden`
10. Check if NFTs that are junk under `Hidden` have context menu with options:
     * Mark as not junk
     * Don't show in wallet
11. Check clicking `Mark as not junk` in step 10 will move NFT to `Collected`, the junk label will be gone
12. Check clicking `Don't show in wallet` in step 10 will remove NFT which you can't see it in neither `Collected` nor `Hidden`
13. Repeat step 5 on the NFT that was moved from `Hidden` to `Collected` from step 11. Now this NFT will be moved to `Hidden` but this time there is no `Junk` label 
14. Try to re-add any deleted NFTs with the same `contractAddress`, network, `symbol` and `tokenId`(if there is one), this NFT will re-appear under `Collected`
## Test Plan:


- Check if a dropdown is available in the header beside filter button in NFT tab
- Check if dropdown selection will display correct group of NFTs
1. Test spam NFTs only come from SimpleHash. This should happen if you never add a custom NFT and manually marked as a spam before. If you have some NFTs are supposed to be auto-discovered but not being displayed under `NFTs` group, they are most likely being displayed under `Spam`
2. Test moving a NFT under `NFTs` to `Spam` by long pressing the NFT then select `Move to Spam`
3. Test hide a NFT under `NFTs` by long pressing the NFT then select `Hide`, this will move NFT to `Hidden`
4. Test unspam a NFT under `Spam` by long pressing the NFT then select `Unspam`, this will move NFT to `NFTs` group
5. Test hide a NFT under `Spam` by long pressing the NFT then select `Hide`, this will move NFT to `Hidden`
6. Test moving a NFT under `Hidden` to `Spam` by long pressing the NFT then select `Move to Spam`
7. Test unhide a NFT under `Hidden` by long pressing the NFT then select `Unhide`, this will move NFT to `NFTs`
- Check no duplication spam NFTs under `Spam`

Test token list:
1. Hide a user asset in Portfolio and make sure this token is from the token registry 

Screens token is hidden | Screens token is not hidden
--- | ---
Portfolio, Send | Token search, Edit visible Asset, Account Details, Swap

1. Add a custom token that it cannot be found from the token registry
2. Hide this custom token in Portfolio 

Screens token is hidden | Screens token is not hidden
--- | ---
Portfolio, Send | Token search, Edit visible Asset, Account Details
Note: transaction summary should also display the token name even though this token is hidden


## Screenshots:

![Simulator Screenshot - iPhone 15 Pro - 2023-10-11 at 16 15 25](https://github.com/brave/brave-ios/assets/1187676/3ab618d0-2364-46ef-a61d-24fbdb3f068b) | ![Simulator Screenshot - iPhone 15 Pro - 2023-10-11 at 16 15 44](https://github.com/brave/brave-ios/assets/1187676/ce0c435c-7b8a-4db5-97c6-61c8e304f469) | ![simulator_screenshot_4B0FB612-2F05-4D11-AEE8-6237B4B12106](https://github.com/brave/brave-ios/assets/1187676/3000bec0-5729-435a-8b05-faa91a8e26b1)
--- | --- | --- 
![Simulator Screenshot - iPhone 15 Pro - 2023-10-11 at 16 15 55](https://github.com/brave/brave-ios/assets/1187676/41a716e2-e6d1-4fc7-9427-9c902eb7bc9e) | ![Simulator Screenshot - iPhone 15 Pro - 2023-10-11 at 16 16 01](https://github.com/brave/brave-ios/assets/1187676/c81bf899-5fb0-44bd-bd28-72d3c54f2d8c) | ![Simulator Screenshot - iPhone 15 Pro - 2023-10-11 at 16 16 06](https://github.com/brave/brave-ios/assets/1187676/60a750d4-8eb2-4b59-b35e-f27efb87bb6e) 


## Reviewer Checklist:

  - `QA/(Yes|No)`
  - `bug` / `enhancement`
